src/helpers/gateway/stall_detection.rs

```diff
@@ -0,0 +1,372 @@
+    };
+
+    impl<M: Message> Observed<crate::helpers::gateway::send::SendingEnd<M>> {
+        delegate::delegate! {
+            to { self.advance(); self.inner() } {
+                #[inline]
+                pub async fn send(&self, record_id: RecordId, msg: M) -> Result<(), Error>;
+            }
+        }
+    }
+
+    pub struct WaitingTasks(BTreeMap<ChannelId, (TotalRecords, Vec<String>)>);
+
+    impl Debug for WaitingTasks {
+        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+            for (channel, (total, records)) in &self.0 {
+                write!(
+                    f,
+                    "\n\"{:?}\", to={:?}. Waiting to send records {:?} out of {total:?}.",
+                    channel.gate, channel.role, records
+                )?;
+            }
+
+            Ok(())
+        }
+    }
+
+    impl ObserveState for GatewaySenders {
+        type State = WaitingTasks;
+
+        fn get_state(&self) -> Option<Self::State> {
+            let mut state = BTreeMap::new();
+            for entry in &self.inner {
+                let channel = entry.key();
+                let sender = entry.value();
+                if let Some(sender_state) = sender.get_state() {
+                    state.insert(channel.clone(), (sender.total_records(), sender_state));
+                }
+            }
+
+            (!state.is_empty()).then_some(WaitingTasks(state))
+        }
+    }
+
+    impl ObserveState for GatewaySender {
+        type State = Vec<String>;
+
+        fn get_state(&self) -> Option<Self::State> {
+            let waiting_indices = self.waiting();
+            super::to_ranges(waiting_indices).get_state()
+        }
+    }
+}
+
+/// Converts a vector of numbers into a vector of ranges.
+/// For example, [1, 2, 3, 4, 5, 7, 9, 10, 11] produces [(1..=5), (7..=7), (9..=11)].
+fn to_ranges(nums: Vec<usize>) -> Vec<std::ops::RangeInclusive<usize>> {
+    nums.into_iter()
+        .fold(Vec::<RangeInclusive<usize>>::new(), |mut ranges, num| {
+            if let Some(last_range) = ranges.last_mut().filter(|r| *r.end() == num - 1) {
+                *last_range = *last_range.start()..=num;
+            } else {
+                ranges.push(num..=num);
+            }
+            ranges
+        })
+}
+
+/// Range formatter that prints one-element wide ranges as single numbers.
+impl<U> ObserveState for Vec<RangeInclusive<U>>
+where
+    U: Copy + Display + Eq + PartialOrd + Ord + Sub<Output = U> + From<u8>,
+{
+    type State = Vec<String>;
+    fn get_state(&self) -> Option<Self::State> {
+        let r = self
+            .iter()
+            .map(
+                |range| match (*range.end() - *range.start()).cmp(&U::from(1)) {
+                    std::cmp::Ordering::Less => format!("{}", range.start()),
+                    std::cmp::Ordering::Equal => format!("[{}, {}]", range.start(), range.end()),
+                    std::cmp::Ordering::Greater => format!("[{}..{}]", range.start(), range.end()),
```

Comment:
It makes the output noisier, but using `..=` here might avoid confusion.

Reply:
I don't mind adding it, I used the math notation where `]` means inclusive and `)` exclusive. Let me know if you think it will still be worth adding `=`